### PR TITLE
feat: Ping (#48)

### PR DESCRIPTION
**Pull request by vrongmeal**
```
$ s2 pingtest my-basin pingtest/express -n 5
Preparing test...
ack: 28 ms    e2e: 29 ms
ack: 28 ms    e2e: 28 ms
ack: 28 ms    e2e: 28 ms
ack: 27 ms    e2e: 27 ms
ack: 28 ms    e2e: 28 ms

Sent 5 batches with 102402 bytes

-------- Append Acknowledgement Latency Statistics ---------
     Mean 28 ms
   Median 28 ms
      P95 28 ms
      P99 28 ms
      Max 28 ms
      Min 27 ms
  Std Dev 0 ms

-------------- End to End Latency Statistics ---------------
     Mean 28 ms
   Median 28 ms
      P95 29 ms
      P99 29 ms
      Max 29 ms
      Min 27 ms
  Std Dev 0 ms
```

Resolves: #48

### CODE REVIEW

**Comment by shikhar** [comment]
please factor out a new mod for as much of the new code as reasonable

**Comment by shikhar** [comment]
@vrongmeal let's just call it `ping` :)

**Comment by vrongmeal** [comment]
Updated:

```
$ s2 ping vr-basin pingtest/express -n 5 -b 50000
Preparing...
50004 bytes:  ack = 30 ms   e2e = 32 ms
50004 bytes:  ack = 29 ms   e2e = 29 ms
49999 bytes:  ack = 28 ms   e2e = 29 ms
49992 bytes:  ack = 29 ms   e2e = 29 ms
50008 bytes:  ack = 29 ms   e2e = 29 ms

Sent 5 batches with 250007 bytes

-------- Append Acknowledgement Latency Statistics ---------
     Mean 29 ms
   Median 29 ms
      P95 30 ms
      P99 30 ms
      Max 30 ms
      Min 28 ms
  Std Dev 0 ms

-------------- End to End Latency Statistics ---------------
     Mean 29 ms
   Median 29 ms
      P95 32 ms
      P99 32 ms
      Max 32 ms
      Min 29 ms
  Std Dev 1 ms
```

File: src/main.rs [hunk 1]
```diff
@@ -830,6 +865,117 @@ async fn run() -> Result<(), S2CliError> {
...
+                interval: Duration,
+                batch_bytes: u64,
+            ) -> Result<(), S2CliError> {
+                let jitter_op = if rand::random() {
+                    u64::saturating_add
+                } else {
+                    u64::saturating_sub
+                };
+
+                let record_bytes = jitter_op(batch_bytes, rand::thread_rng().gen_range(0..10));
```
Comment (shikhar) [inline src/main.rs:901]: the jitter seems too small, you can do like +/- 50% of the target size

Comment (vrongmeal) [reply src/main.rs:901]: So 20 KiB +/- 10 KiB ? Isn't that too big?

File: src/main.rs [hunk 2]
```diff
@@ -830,6 +865,117 @@ async fn run() -> Result<(), S2CliError> {
...
+
+            let total_batches = pings.len();
+            let (bytes, (acks, e2es)): (Vec<_>, (Vec<_>, Vec<_>)) = pings
+                .into_iter()
+                .map(|PingResult { bytes, ack, e2e }| (bytes, (ack, e2e)))
+                .unzip();
+            let total_bytes = bytes.into_iter().sum::<u64>();
+
+            eprintln!(/* Empty line */);
+            eprintln!("Sent {} batches with {} bytes", total_batches, total_bytes);
```
Comment (shikhar) [inline src/main.rs:938]: `Sent {total_bytes} bytes in {total_batches} batches`

File: src/main.rs (hunk 1 above)
Comment (shikhar) [reply src/main.rs:901]: Hmm maybe 25%... want to exercise a wider range of variation. You can also grow the default to 32KiB and max to 128KiB.

File: src/ping.rs [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    reads_rx: mpsc::UnboundedReceiver<Result<(Instant, SequencedRecord), S2CliError>>,
+}
+
+impl Pinger {
+    pub async fn init(stream_client: &StreamService) -> Result<Self, S2CliError> {
+        let tail = stream_client.check_tail().await?;
+
+        let mut read_stream = stream_client.read_session(tail, None, None).await?;
+
+        let (stream_tx, stream_rx) = mpsc::unbounded_channel();
```
Comment (shikhar) [inline src/ping.rs:36]: `stream` gets over-used in our contexts, maybe do `records_{tx,rx}`

File: src/ping.rs [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        let mut append_stream = stream_client
+            .append_session(
+                tokio_stream::wrappers::UnboundedReceiverStream::new(stream_rx),
+                AppendRecordsBatchingOpts::new()
+                    .with_max_batch_records(1)
+                    .with_match_seq_num(Some(tail)),
+            )
+            .await?;
+
+        let warm_up_record = AppendRecord::new("warm up").expect("valid record");
```
Comment (shikhar) [inline src/ping.rs:46]: serious nit :P `warmup` is one word

**Review by shikhar — APPROVED** [review]
looks great!

**Pull request merged**